transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+}
+
+impl GiftCodeMemoBuilder {
+    /// Create a new gift code memo builder specifying the memo type
+    pub fn new(gift_code_memo_type: GiftCodeMemoType) -> Self {
+        Self {
+            gift_code_memo_type,
+            note: "",
+            gift_code_tx_out_public_key: None,
+            gift_code_tx_out_global_index: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+        }
+    }
+    /// A note in the form of null terminated utf-8 bytes can be applied to
+    /// gift code funding or sender memos. These notes end up in the change
+    /// outputs and indicate what the memo was for. Set this memo here. This
+    /// method will enforce the 60 and 64 byte length for Funding Memos and
+    /// Sender Memos respectively. If the memo type is a cancellation memo
+    /// and a note is set, any value set here will be ignored as the
+    /// cancellation memo does not include a note.
+    pub fn set_note(&mut self, note: &'static str) -> Result<(), MemoError> {
+        let max_note_length = if let GiftCodeMemoType::Funding = self.gift_code_memo_type {
+            60
+        } else {
+            64
+        };
+        if note.len() > max_note_length {
+            return Err(MemoError::BadLength(note.len()));
+        }
+        self.note = note;
+        Ok(())
+    }
+    /// Clear the gift code note
+    pub fn clear_note(&mut self) {
+        self.note = "";
+    }
+    /// Set tx_out_public_key to be written to a sender memo. If the type is not
+    /// a sender memo the key isn't written.
+    pub fn set_gift_code_tx_out_public_key(&mut self, tx_out_public_key: RistrettoPublic) {
+        if let GiftCodeMemoType::Sender = self.gift_code_memo_type {
+            self.gift_code_tx_out_public_key = Some(tx_out_public_key);
+        }
+    }
+    /// Clear the gift code tx_out_public_key
+    pub fn clear_gift_code_tx_out_public_key(&mut self) {
+        self.gift_code_tx_out_public_key = None;
+    }
+    /// Set global index for the TxOut
+    pub fn set_gift_code_tx_out_index(&mut self, tx_out_global_index: u64) {
+        self.gift_code_tx_out_global_index = Some(tx_out_global_index)
+    }
+    /// Clear global index for gift code TxOut
+    pub fn clear_gift_code_tx_out_index(&mut self) {
+        self.gift_code_tx_out_global_index = None;
+    }
+    /// Enable gift code change memos
+    pub fn enable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = true;
+    }
+    /// Disable gift code change memos
+    pub fn disable_gift_code_change_memo(&mut self) {
+        self.gift_code_change_memo_enabled = false;
+    }
+}
+
+impl MemoBuilder for GiftCodeMemoBuilder {
+    /// Set the fee
+    fn set_fee(&mut self, _fee: Amount) -> Result<(), NewMemoError> {
+        Ok(())
+    }
+
+    /// Gift code destination memos are not allowed - RTH transaction builders
+    /// should be used alongside this builder to create recoverable
```

Comment:
How would that work given that `TransactionBuilder` can only accept a single `MemoBuilder`?

Reply:
This is is the design consideration I'm hoping to discuss as it affects the way future memo builders would work. The MCIP as written specifies we should be able to write destination memos alongside recoverable transaction history.

However, as you point out, it can only take ONE memo builder at the moment - so IMHO - the considerations are:
1. Do we want RTH to work as normal when generating special transaction/memo types?
- If we RTH to work as normal alongside other memo types (like this one), that means we build the capability to use two memo builders alongside each other OR the memo builder for special transaction type (burn, gift code, etc.) duplicates the work of RTH, which is kind of wasteful.

2. Are okay with RTH not working exactly the same for special memo types?
- If we're fine with this, then we can use a single memo builder no issue

Personally I think RTH working as normal for all/most transaction/memo types would keep the most consistent accounting if we somehow want to audit. However, it may not be necessary. What are thoughts?

Reply:
I don't think this is aligned with my understanding of what RTH is. According to https://github.com/mobilecoinfoundation/mcips/blob/main/text/0032-fog-compatible-gift-codes.md#recoverable-transaction-history-and-memos, when creating gift codes we should be writing a memo for both the change output (`0x0201 Gift code funding memo`) and the actual gift code output (`0x0002 Gift code sender memo`).
It appears that the former is handled in this PR in `make_memo_for_change_output`, but the latter is not in `make_memo_for_output`.

I think when someone wants to send out a gift code, they will be creating a TransactionBuilder with the `GiftCodeMemoBuilder` (or perhaps it should be much more specific - a `GiftCodeFundingMemoBuilder`). They then add the single output that is the actual gift code output. There's no need for the "standard" (https://github.com/mobilecoinfoundation/mobilecoin/blob/b6af448d1ff5f980f05ac641aa087a99744d0e96/transaction/std/src/memo_builder/rth_memo_builder.rs) memos in this case, if I understood everything correctly.

The one thing I don't like here is that the user of the `TransactionBuilder` needs to be aware of the specific steps they need to take in order to generate a valid gift code, namely constructing the `TransactionBuilder` with the right memo builder, and calling `add_output` only once (which should be enforced by the memo builder) with the correct destination (the gift code subaddress).

It might be useful to provide a `GiftCodeFundingBuilder` object that embeds all this logic in it.

Either way, I'd like to see a unit test that uses the gift code memo builder(s) (might be worth splitting into twodifferent builders since the behavior is actually very different for the scenarios of funding and cancellation) with the `TransactionBuilder` to produce actual `Tx`s that contain the expected set of outputs and memos.

Reply:
It seems to make sense that the 0x0002 goes to the gift code output but it actually is specified to go to the change output. A little strange, but that's what's in there. We could potentially revisit that in the MCIP.

> When claiming a gift code, the gift code output is spent in a self-payment to the change subaddress. This transaction has a single output, the change output, which has a memo of type 0x0002 Gift code sender memo. 

Also noted about the tx builder user having having to be aware of what state they're at potentially causing problems - I will think about this.

Also thanks for the design input. I do intend to build tests, but I wanted to first sanity check the initial architecture with the hive mind before I started building too heavily in a specific direction. I will go ahead and address all the feedback here and add thorough tests.

Reply:
I see, interesting, I misunderstood that about `0x0002` but it does make sense.
So the output that goes to the gift code subaddress does not actually have a memo?

In that case it seems like funding a giftcode should allow for two outputs (change + regular output), with the regular one not having a memo and canceling/claiming a gift code should only allow for a change output. Is that correct?

Reply:
> So the output that goes to the gift code subaddress does not actually have a memo?

Yeah as written that's correct. It could potentially have a different design, but we'd probably want to get larger stakeholder buy in about changing the MCIP.

> In that case it seems like funding a giftcode should allow for two outputs (change + regular output), with the regular one not having a memo and canceling/claiming a gift code should only allow for a change output. Is that correct?

That is correct!